[13.0][FIX] product: still missing attributes and relations

MiquelRForgeFlow opened the pull request:
Explained in code, see each commit. It's a continuation of https://github.com/OCA/OpenUpgrade/pull/2633.




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

pedrobaeza commented:
I don't see enough explanations neither on the PR message nor the commit messages. Please do the same as I did in #2633.

MiquelRForgeFlow commented:
Yeah, yesterday I did it a bit fast.

The second commit is easy to explain. After migration of a client that has mrp, there were some bom lines that had lost their attribute, just the same way/problem you had previously with products that you solved previously. And example for this case should be a copy pasta of the ones you put but changing products by bom lines. And that's why the queries are very similar.

About the first commit it's a bit tricky. It happens that the new archived attributes you created, after migration, they still had null attribute_line_id, which is a required field. The problem lies in that when you created the archived attribute lines, you should also fill the m2m table between PAV and PTAL, and that's what the commit basically does. It also includes two things: 1) in postmigration, first check for active ptals and then for inactive ptals, always trying to find ptals because we want all attribute_line_id filled. 2) in pre-migration, we begin to fill attribute_line_id for the new archived ptavs (using the previous m2m table between PAV and PTAL), and thus in post-migration this ones won't need have attribute_line_id filled.

MiquelRForgeFlow committed:
[OU-IMP] product: make us of index for better performance

MiquelRForgeFlow committed:
[OU-IMP] product: simplify combination_indices query

More readable and no need to provision the whole ungrouped table, just
the group results.

MiquelRForgeFlow committed:
[OU-IMP] product: fill attribute_line_id of ptav in pre-migration

It's needed for other later queries to be pre-filled, so it's done in
pre-migration and as soon as possible.

MiquelRForgeFlow committed:
[OU-FIX+IMP] product: Populate all required ptav + fill attribute_line_id

An extra link requirement is added on the LEFT JOIN that checks if the
proper product_template_attribute_value exists. If not, there can be
false matches that prevents the creation of the line.

Steps to reproduce (that partially fixes this commit):

- Create a Product Template with:

  Attribute lines: Color (Red, Blue), Material (Iron, Wood)

- Create and Validate transfer with Variant (Red, Iron)
- Remove Red value from attribute line
  (on this step we lose here link between product.attribute.value and product.template.attribute.value)
- run migration
- Go to Product Template form
- Press on Configure Variants button
 -Set Inactive filter

Current result: Missing attribute for inactive product.template.attribute.value.

We also fill at the same time the attribute_line_id values for this
newly created records, for not having to be filled later.

MiquelRForgeFlow committed:
[OU-FIX] product: fill attribute_id of ptavs in pre-migration

When the module is loading, the new product_attribute_product_template_rel
table will be automatically created and filled (it's an stored computed m2m).
Thus, the attribute_id of ptavs is needed in such compute, and we need
to move the filling of the attribute_id of ptavs to pre-migration.

MiquelRForgeFlow committed:
[OU-IMP] product: make us of openupgrade.add_fields

MiquelRForgeFlow committed:
[OU-FIX] product: make use of mrp_bom_line_product_attribute_value_rel

We have to fix similar case of missing archived ptals and ptavs in mrp,
due to bom lines missing those values (which in v13 are archived instead
of deleted). We do it here instead of mrp because we want those archived
ptals and ptavs have their new fields and m2m table relations filled.

Steps to reproduce:

- Create a Product Template with:

  Attribute lines: Color (Red, Blue), Material (Iron, Wood)

- Create BOM for this Product in the BOM line set apply on Color: Red.
- Remove Red value from attribute line.
  (on this step product.template.attribute.value and product.variant will be removed)
- run migration.
- Go to Product Template form.
- Press on Configure Variants button.
- Set Inactive filter not found.

Current Result: Missing Color: Red record.

- Go to BOM form.

Current Result: Missing Color: Red in BOM line

MiquelRForgeFlow commented:
> I don't see enough explanations neither on the PR message nor the commit messages

Done. Now you can check commit by commit what each minor change is about.

pedrobaeza commented on addons/product/migrations/13.0.1.2/pre-migration.py:
```diff
@@ -174,7 +174,7 @@ def migrate(env, version):
+    calculate_product_product_combination_indices(env)
     insert_missing_product_template_attribute_line(env)
     insert_missing_product_template_attribute_value(env)
-    calculate_product_product_combination_indices(env)
```
If the ptvas were archived on v13, they has combination_indices, isn't it? Why not filling such ones then?

MiquelRForgeFlow replied:
Ok, you have a point here. I don't know why but I thought archived ptavs were avoided. Dropped that commit.

pedrobaeza commented on addons/product/migrations/13.0.1.2/openupgrade_analysis_work.txt:
```diff
@@ -83,7 +83,7 @@ product      / product.template.attribute.line / product_template_value_ids (man
 product      / product.template.attribute.line / product_template_value_ids (many2many): type is now 'one2many' ('many2many')
 product      / product.template.attribute.value / attribute_line_id (many2one)  : NEW relation: product.template.attribute.line, required
-# DONE: post-migration: filled field (using the same approach than in OU v12)
+# DONE: pre-migration: filled required field (using the same approach than in OU v12)
```
If I don't remember bad, if you this, there are missing values that are fill lately.

MiquelRForgeFlow replied:
I initially moved that to pre-migration because the field attribute_line_id of ptav is required. Later, it proves useful in pre-migration, in correctly filling the missing ptavs (the query of missing ptavs use that field in order to make it compatible with the sql_constraint `unique(attribute_line_id, product_attribute_value_id)`)

pedrobaeza commented on addons/product/migrations/13.0.1.2/pre-migration.py:
```diff
@@ -125,18 +125,32 @@ def insert_missing_product_template_attribute_value(env):
+        "ALTER TABLE product_template_attribute_value "
+        "ADD COLUMN attribute_line_id INT4",
+    )
+    fill_product_template_attribute_value_attribute_line_id(env)
```
Please don't nest this high level methods.

MiquelRForgeFlow replied:
Ok, done.

pedrobaeza commented on addons/product/migrations/13.0.1.2/pre-migration.py:
```diff
@@ -88,6 +88,39 @@ def insert_missing_product_template_attribute_line(env):
         AND insert_ptal.attribute_id = pav.attribute_id)
     GROUP BY pav.id, insert_ptal.id""",
     )
+    if openupgrade.table_exists(env.cr, "mrp_bom_line"):
```
I don't get the relation between product_template_attribute_line and mrp boms, neither the code has any reference to this.

MiquelRForgeFlow replied:
It's explained in that commit message. In mrp module, the bom lines also have attributes, and some are missing in a similar way the missing attributes on products case you detected. In order to avoid these missing attributes being missing, we need to make use of mrp_bom_line_product_attribute_value_rel, which plays a very similar role that product_attribute_value_product_product_rel has for products. We have tested it and works excellent.

pedrobaeza reviewed (changes requested):
I still miss an explanation of each change about why, and enumerating the steps to reproduce the problem. Please take as example https://github.com/OCA/OpenUpgrade/pull/2633, where the exact initial data and the expected ones are shown.

Without that, everything is done blindly, and the SQL modifications are very convoluted for validating them reading them.

MiquelRForgeFlow commented:
> I still miss an explanation of each change about why, and enumerating the steps to reproduce the problem

Well, I thought I added the enough explanations in the commit message, in the code or in the work.txt (depending the commit). Just tell me exactly which commits you need more info about and I will try to add more explanation on them.

MiquelRForgeFlow committed:
[OU-FIX] product,mrp: consider ptav.attribute_id in some queries

MiquelRForgeFlow commented:
PR updated :)

kos94ok-3D commented on addons/product/migrations/13.0.1.2/pre-migration.py:
```diff
@@ -134,38 +171,76 @@ def calculate_product_product_combination_indices(env):
+        "ALTER TABLE product_template_attribute_value "
+        "ADD COLUMN attribute_line_id INT4",
+    )
+    # use ptal.active = TRUE
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE product_template_attribute_value ptav
+        SET attribute_line_id = ptal.id
+        FROM product_template_attribute_line ptal
+        JOIN product_template pt ON ptal.product_tmpl_id = pt.id
+        JOIN product_attribute_value pav ON ptal.attribute_id = pav.attribute_id
+        WHERE ptal.active = TRUE AND ptav.product_tmpl_id = pt.id
+            AND ptav.product_attribute_value_id = pav.id
+            AND ptav.attribute_line_id IS NULL""",
+    )
+    # use ptal.active != TRUE
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE product_template_attribute_value ptav
+        SET attribute_line_id = ptal.id
+        FROM product_template_attribute_line ptal
+        JOIN product_template pt ON ptal.product_tmpl_id = pt.id
+        JOIN product_attribute_value pav ON ptal.attribute_id = pav.attribute_id
+        WHERE ptal.active IS DISTINCT FROM TRUE AND
+            ptav.product_tmpl_id = pt.id AND
+            ptav.product_attribute_value_id = pav.id
+            AND ptav.attribute_line_id IS NULL""",
+    )
```
Now we can do it in one query:
<pre>
  openupgrade.logged_query(
      env.cr, """
      UPDATE product_template_attribute_value ptav
      SET attribute_line_id = ptal.id
      FROM product_template_attribute_line ptal
      JOIN product_attribute_value pav ON ptal.attribute_id = pav.attribute_id
      WHERE ptav.product_tmpl_id = ptal.product_tmpl_id
          AND ptav.product_attribute_value_id = pav.id
          AND ptav.attribute_line_id IS NULL""",
  )
</pre>

kos94ok-3D commented on addons/mrp/migrations/13.0.2.0/post-migration.py:
```diff
@@ -14,12 +15,14 @@ def fill_bom_product_template_attribute_value(env):
+    JOIN product_attribute_value pav ON pav.id = mbl_pav_rel.product_attribute_value_id
     JOIN mrp_bom_line mbl ON mbl.id = mbl_pav_rel.mrp_bom_line_id
     JOIN mrp_bom mb ON mbl.bom_id = mb.id
     JOIN product_template pt ON mb.product_tmpl_id = pt.id
```
useless join we need only `id` which we already have in `mb.product_tmpl_id`

MiquelRForgeFlow replied:
that's https://github.com/OCA/OpenUpgrade/pull/2957 😆

kos94ok-3D commented on addons/product/migrations/13.0.1.2/pre-migration.py:
```diff
@@ -63,20 +58,37 @@ def insert_missing_product_template_attribute_line(env):
+            FROM mrp_bom_line_product_attribute_value_rel mblpavr
+            JOIN mrp_bom_line mbl ON mbl.id = mblpavr.mrp_bom_line_id
+            JOIN mrp_bom mb ON mbl.bom_id = mb.id
+            JOIN product_template pt ON pt.id = mb.product_tmpl_id
```
useless join we need only `id` which we already have in `mb.product_tmpl_id`

kos94ok-3D commented on addons/product/migrations/13.0.1.2/pre-migration.py:
```diff
@@ -107,20 +119,45 @@ def insert_missing_product_template_attribute_value(env):
+            FROM mrp_bom_line_product_attribute_value_rel mblpavr
+            JOIN mrp_bom_line mbl ON mbl.id = mblpavr.mrp_bom_line_id
+            JOIN mrp_bom mb ON mbl.bom_id = mb.id
+            JOIN product_template pt ON pt.id = mb.product_tmpl_id
```
useless join we need only `id` which we already have in `mb.product_tmpl_id`

kos94ok-3D reviewed (commented):
Work as expected.
I have some cosmetic suggests

kos94ok-3D commented:
@MiquelRForgeFlow You forgot about `group by` section

MiquelRForgeFlow commented:
@kos94ok-3D done

kos94ok-3D commented:
HI, @MiquelRForgeFlow
Are we waiting for the review from @pedrobaeza?

MiquelRForgeFlow commented:
> Are we waiting for the review from @pedrobaeza?

Yeah, a long awaited review. He wants me to wait for his reviews in this repo for all the PRs.

pedrobaeza commented:
I still need commit messages explaining the whys of these commits:

97142fbe68e913dc245225828c533887c2326060
c263c885cb9ff813d7336046b4af3f2a7e535ee2
e54a1412c6c2710973a4783eb148cafbb8759e20
1bc8c9820c6dbec4e277cdc00114579264da409e
f6dbddc5400cc22fbf084e743cc7f42c9e5eeac0
a823cdec234b7b6902349f147b6f858f987cafb9

and steps to reproduce a problem.

Please see a good example like the one I did for a similar case: 3dabbccb7d96c9627fae0f1e4fbdf790c1589be9

MiquelRForgeFlow commented:
- https://github.com/OCA/OpenUpgrade/commit/97142fbe68e913dc245225828c533887c2326060 - it's just a simplification. It reduces the selects by one.
- https://github.com/OCA/OpenUpgrade/commit/e54a1412c6c2710973a4783eb148cafbb8759e20 is the core of this PR. All this refactor began with this commit, which fixes the correct number of archived ptavs that are being created. The key is adding this condition in the left join:
![Selection_054](https://user-images.githubusercontent.com/25005517/141996834-b0a1f0b6-f3cb-4167-aeac-a08a17b9abe7.png)
And to add this condition, then we need to have that `ptav.attribune_line_id` filled, which is done previously in https://github.com/OCA/OpenUpgrade/commit/c263c885cb9ff813d7336046b4af3f2a7e535ee2.
- https://github.com/OCA/OpenUpgrade/commit/1bc8c9820c6dbec4e277cdc00114579264da409e already has the explanation included in the docstring (of `fill_product_template_attribute_value__attribute_id_related` method).
- https://github.com/OCA/OpenUpgrade/commit/f6dbddc5400cc22fbf084e743cc7f42c9e5eeac0 just solves a similar case of losing data (and thus not having enough archived ptavs). The mrp_bom_line_product_attribute_value_rel table becomes deprecated and is replaced by mrp_bom_line_product_template_attribute_value_rel in mrp module. This extra archived ptavs are created here instead of doing it in mrp module because we can take advantage of the other queries to fill the collateral data. If not, in mrp we would have to copy and paste the same queries, like the one that fills attribute_id in ptavs, and thus duplicating a lot of code.
- https://github.com/OCA/OpenUpgrade/commit/a823cdec234b7b6902349f147b6f858f987cafb9 this one is a pretty obvious fix and I doubt it needs any explanation. Again, without this extra conditions, we are having wrong data.

kos94ok-3D commented:
Case1 Steps to reproduce:
1. Create a Product Template with
   Attribute lines: `Color (Red, Blue)`, `Material (Iron, Wood)`
 2. Create and Validate transfer with Variant `(Red, Iron)`
 3. Remove `Red` value from attribute line
  (on this step we lose here link between `product.attribute.value` and `product.template.attribute.value`)
 4. run migration
 5. Go to Product Template form
 6. Press on `Configure Variants` button 
 7. Set `Inactive` filter
 
Actual Result:
Missing attribute for inactive `product.template.attribute.value`

Case 2 Steps to reproduce:
1. Create a Product Template with
   Attribute lines: `Color (Red, Blue)`, `Material (Iron, Wood)`
 2. Create BOM for this Product in the BOM line set apply on `Color: Red`
 3. Remove `Red` value from attribute line
  (on this step `product.template.attribute.value` and `product.variant` will be removed)
 4. run migration
 5. Go to Product Template form
 6. Press on `Configure Variants` button 
 7. Set `Inactive` filter not found 
Actual Result:
Missing `Color: Red` record

8. Go to BOM form
Actual Result:
Missing `Color: Red` in BOM line

pedrobaeza reviewed (approved):
Pushed a rebased branch with all of these comments in commit messages and review thoroughly. There are few doubts about some extra linking, but let's see applied to real migrations.
